File: JAVA/200. Number of Islands.java
class Solution {
        public int numIslands(char[][] grid) {
            int h = grid.length;   
            if (h == 0) return 0;  
            int l = grid[0].length; 
            int result = 0;
            for (int i = 0; i < h; i++) {  //loop iterates over rows
                for (int j = 0; j < l; j++) {
                    if (grid[i][j]=='1'){
                        dfs(grid, i, j);
                        result++;
                    }
                }
            }
            return result;
        }
        public void dfs(char[][] grid, int row, int col){
            int H = grid.length;
            int L = grid[0].length;
            if (row < 0 || col < 0 || row >= H || col >= L || grid[row][col] != '1'){
                return;
            }
            grid[row][col]='0';  
            dfs(grid, row+1, col); 
            dfs(grid, row-1, col);  
            dfs(grid, row, col+1);  
            dfs(grid, row, col-1);  
        }
}

File: C++/200. Number of Islands.cpp
class Solution {
public:
    void bfs(vector<vector<char>> &grid, vector<vector<int>> &visited, int row, int col){
        queue<pair<int, int>> q;
        q.push({row, col});
        visited[row][col] = 1;
        while(!q.empty()){
            int i = q.front().first;
            int j = q.front().second;
            q.pop();
            if(i + 1 < grid.size() && grid[i+1][j] == '1' && !visited[i+1][j]){
                visited[i+1][j] = 1;
                q.push({i+1, j});
            }
            if(j + 1 < grid[0].size() && grid[i][j+1] == '1' && !visited[i][j+1]){
                visited[i][j+1] = 1;
                q.push({i, j+1});
            }
            if(i - 1 >= 0 && grid[i-1][j] == '1' && !visited[i-1][j]){
                visited[i-1][j] = 1;
                q.push({i-1, j});
            }
            if(j - 1 >= 0 && grid[i][j-1] == '1' && !visited[i][j-1]){
                visited[i][j-1] = 1;
                q.push({i, j-1});
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        vector<vector<int>> visited(grid.size(), vector<int>(grid[0].size(), 0));
        for(int i = 0; i<grid.size(); i++){
            for(int j = 0; j<grid[0].size(); j++){
                if(grid[i][j] == '1' && !visited[i][j]){
                    count++;
                    bfs(grid, visited, i , j);
                }
            }
        }
        return count;
    }
};
